libupload.groovy

```diff
@@ -1,4 +1,85 @@
 
+// Replicate artifacts in various clouds
+def replicate_to_clouds(pipecfg, basearch, buildID) {
+
+    def meta = readJSON(text: shwrapCapture("cosa meta --arch=${basearch} --dump"))
+    def replicators = [:]
+    def credentials
+
+    credentials = [file(variable: "ALIYUN_IMAGE_UPLOAD_CONFIG",
+                        credentialsId: "aliyun-image-upload-config")]
+    if (meta.aliyun) {
+        def creds = credentials
+        replicators["☁️ 🔄:aliyun"] = {
+            withCredentials(creds) {
+                utils.syncCredentialsIfInRemoteSession(["ALIYUN_IMAGE_UPLOAD_CONFIG"])
+                def c = pipecfg.clouds.aliyun
+                def extraArgs = []
+                if (c.public) {
+                    extraArgs += "--public"
+                }
+                shwrap("""
+                coreos-assembler aliyun-replicate \
+                    --build=${buildID} \
+                    --arch=${basearch} \
+                    --config=\${ALIYUN_IMAGE_UPLOAD_CONFIG} \
+                    ${extraArgs.join(' ')}
+                """)
+            }
+        }
+    }
+    if (meta.amis) {
+        replicators["☁️ 🔄:aws"] = {
+            def replicated = false
+            def ids = ['aws-build-upload-config', 'aws-govcloud-image-upload-config']
+            def c = pipecfg.clouds.aws
+            for (id in ids) {
+                tryWithCredentials([file(variable: 'AWS_CONFIG_FILE', credentialsId: id)]) {
+                    shwrap("""
+                    cosa aws-replicate \
+                        --build=${buildID} \
+                        --arch=${basearch} \
+                        --source-region ${c.primary_region} \
```

Comment:
we can delete `--source-region` here. The code will autodetect it from the `meta.json` if not specified.